File: answers/answers-rb/tree-search.rb
require 'json'
require 'active_support/all'

path_relative_base = '../../questions/'
path =
  File.join File.dirname(__FILE__), path_relative_base, 'tree-search/tree.rb'

j = JSON.parse `ruby #{path} gen dan`

def process(node, sentence = "", score = 0)
  new_sentence = (sentence + ' ' + node["name"]).gsub(/\s+/, ' ').strip
  new_score = score + node["score"]
  if node["children"].empty?
    [{ sentence: new_sentence, score: new_score }]
  else
    node["children"].map { |child| process(child, new_sentence, new_score)}.flatten
  end
end

results = process j
max_score = results.max_by { |r| r[:score] }[:score]
puts results.select { |r| r[:score] >= max_score }

File: answers/answers-js/tree-search/tree-search.js
//note this is using the username "javascript"
//ruby tree.rb gen javascript > data.json
let tree = require("./data.json")
let bestSentences = ""
let bestScore = 0
//recursively find all possible paths through the tree
const findPaths = (node, sentence, score)=>{
  let paths = []
  sentence += node.name+" "
  score += node.score
  if (node.children.length>0) {
    for (let child of node.children) {
      findPaths(child, sentence, score)
    }
  } else {
    if (score>bestScore) {
        bestScore = score
        bestSentence = sentence
    }
  }
}
let sentences = findPaths(tree,"", 0)
console.log(bestScore, bestSentence)